src/CommandResourceEstimator.cpp

```diff
@@ -0,0 +1,126 @@
+#include "CommandResourceEstimator.hpp"
+#include "CommandRunner.hpp"
+#include "Helpers.hpp"
+
+#include <google/protobuf/util/json_util.h>
+
+#include <process/defer.hpp>
+#include <process/dispatch.hpp>
+
+namespace criteo {
+namespace mesos {
+
+using std::string;
+
+using ::mesos::modules::Module;
+using ::mesos::Resource;
+using ::mesos::Resources;
+using ::mesos::slave::ResourceEstimator;
+using ::mesos::ResourceUsage;
+
+using process::Future;
+using process::Failure;
+using process::terminate;
+
+class CommandResourceEstimatorProcess
+    : public process::Process<CommandResourceEstimatorProcess> {
+ public:
+  CommandResourceEstimatorProcess(
+      const lambda::function<Future<ResourceUsage>()>& usage,
+      const Option<Command>& oversubscribableCommand, bool isDebugMode);
+
+  virtual Future<Resources> oversubscribable();
+  virtual Resources m_oversubscribable(const ResourceUsage& usage);
+  inline const Option<Command>& oversubscribableCommand() const {
+    return m_oversubscribableCommand;
+  }
+
+ protected:
+  const lambda::function<Future<ResourceUsage>()> m_usage;
+
+ private:
+  const Option<Command> m_oversubscribableCommand;
+  bool m_isDebugMode;
+};
+
+CommandResourceEstimatorProcess::CommandResourceEstimatorProcess(
+    const lambda::function<process::Future<ResourceUsage>()>& usage,
+    const Option<Command>& oversubscribableCommand, bool isDebugMode)
+    : m_usage(usage),
+      m_oversubscribableCommand(oversubscribableCommand),
+      m_isDebugMode(isDebugMode) {}
+
+Resources CommandResourceEstimatorProcess::m_oversubscribable(
+    const ResourceUsage& m_usage) {
+  // Mocking a resources to have valid return
+  Try<Resources> l_resources = ::mesos::Resources::parse(
+      "[{\"name\" : \"cpus\", \"type\":\"SCALAR\", \"scalar\" : {\"value\" : "
+      "\"0\"}, \"role\" : \"*\"}]");
+  Resources resources = l_resources.get();
+
+  if (m_oversubscribableCommand.isNone()) {
+    return resources;
+  }
+  logging::Metadata metadata = {"0", "oversubscribable"};
+
+  CommandRunner cmdrunner = CommandRunner(m_isDebugMode, metadata);
+
+  string input;
+  google::protobuf::util::MessageToJsonString(m_usage, &input);
+
+  Try<string> output = cmdrunner.run(m_oversubscribableCommand.get(), input);
+  if (output.isError()) {
+    LOG(INFO) << output.error();
+    // return Failure("output from CommandRunner gave error: " +
+    // output.error());
+    return resources;
```

Comment:
it sounds dangerous to return "0 cpus"  whenever the command fail. Maybe we could:
- store last result from the script (if any)
- otherwise, explicitely fail

Reply:
Actually i think returning **previous result** is more **dangerous** since if it keep failing it will always have the same amount thus **allocating more revocable task that the Agent** could take on.
Returning 0 prevent any revocable task to be allocated. In my tough if the Resource Estimator is broken revocable task should **not be able to allocate**.
I did try to fail, since the resource are returned to the slave it does not take fail from oversubscribable command, apart from modifying the slave i am not sure how we could make it work, if you have any suggestion i am interested